File: backend/app/src/main/kotlin/net/animeta/backend/controller/v2/SearchController.kt
package net.animeta.backend.controller.v2

import net.animeta.backend.service.SearchService
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/v2/search")
class SearchController(private val searchService: SearchService) {
    data class SearchResultItem(val title: String, val n: Int, val id: Int)

    @GetMapping
    fun search(@RequestParam("q") query: String,
               @RequestParam("min_record_count", defaultValue = "2") minRecordCount: Int): List<SearchResultItem> {
        return searchService.searchWorks(query, limit = 30, minRecordCount = minRecordCount)
                .map { SearchResultItem(it.title, it.record_count, it.workId) }
    }

    @GetMapping("/suggest")
    fun suggest(@RequestParam("q") query: String): List<SearchResultItem> {
        return searchService.suggestWorks(query, limit = 30)
                .map { SearchResultItem(it.title, it.record_count, it.workId) }
    }
}

File: backend-ts/src/controllers/search.controller.ts
import { Controller, Get, Query } from "@nestjs/common";
import { SearchService } from "src/services/search.service";
import { SearchResultItem } from 'shared/types';

@Controller('/api/v4/search')
export class SearchController {
  constructor(private searchService: SearchService) {}

  @Get()
  async search(@Query('q') q: string): Promise<SearchResultItem[]> {
    return await this.searchService.search(q, 30)
  }

  @Get('suggest')
  async suggest(@Query('q') q: string): Promise<SearchResultItem[]> {
    return await this.searchService.suggest(q, 30)
  }
}
